### PR TITLE
network: make Type=ether match based on iftype

### LINKED ISSUE #14952: systemd.network(5) Type=ether does not match due to DEVTYPE not being set by udev
**systemd version the issue has been seen with**
> systemd 244 (244.3-1-arch)

**Used distribution**
 > Arch Linux

**Expected behaviour you didn't see**
 > Need to create a default network configuration for all Ethernet devices (similar to #3998).

**Unexpected behaviour you saw**
 > `systemd.network(5)` files using `Type=ether` do not match Ethernet devices.

**Steps to reproduce the problem**

```shell
cat<<EOF>/etc/systemd/network/00-ether-test.network
[Match]
Type=ether
[Network]
DHCP=ipv4
EOF

networkctl reload
```

# Summary

When specifying `Type=ether` in `systemd.network(5)`, Ethernet devices are not matched. I believe this is because `udev(7)` does not define a `DEVTYPE`  property for Ethernet devices. However, when running `networkctl status`, `networkctl(1)` correctly identifies the device as `Type: ether`.

## How the issue was verified

Quote from `systemd.network(5)`:
>**Type=**
>           A whitespace-separated list of shell-style globs matching
>           the device type, as exposed by the udev property
>           "DEVTYPE". If the list is prefixed with a "!", the test
>           is inverted.

`udevadm info` shows that no `DEVTYPE` property is set.

```console
# udevadm info /sys/class/net/enp0s31f6
P: /devices/pci0000:00/0000:00:1f.6/net/enp0s31f6
L: 0
E: DEVPATH=/devices/pci0000:00/0000:00:1f.6/net/enp0s31f6
E: INTERFACE=enp0s31f6
E: IFINDEX=2
E: SUBSYSTEM=net
E: USEC_INITIALIZED=22963369
E: ID_NET_NAMING_SCHEME=v243
E: ID_NET_NAME_MAC=enx482ae3fa5e16
E: ID_OUI_FROM_DATABASE=Wistron InfoComm(Kunshan)Co.,Ltd.
E: ID_NET_NAME_PATH=enp0s31f6
E: ID_BUS=pci
E: ID_VENDOR_ID=0x8086
E: ID_MODEL_ID=0x15bc
E: ID_PCI_CLASS_FROM_DATABASE=Network controller
E: ID_PCI_SUBCLASS_FROM_DATABASE=Ethernet controller
E: ID_VENDOR_FROM_DATABASE=Intel Corporation
E: ID_MODEL_FROM_DATABASE=Ethernet Connection (7) I219-V
E: ID_PATH=pci-0000:00:1f.6
E: ID_PATH_TAG=pci-0000_00_1f_6
E: ID_NET_DRIVER=e1000e
E: ID_NET_LINK_FILE=/usr/lib/systemd/network/99-default.link
E: SYSTEMD_ALIAS=/sys/subsystem/net/devices/enp0s31f6
E: TAGS=:systemd:
```

`networkctl status` shows `Type: ether`, however, no `Network File:` is matching  with the above `systemd.network(5)` file.

```console
# networkctl status enp0s31f6
● 2: enp0s31f6
             Link File: /usr/lib/systemd/network/99-default.link
          Network File: n/a
                  Type: ether
                 State: routable (unmanaged)
                  Path: pci-0000:00:1f.6
                Driver: e1000e
                Vendor: Intel Corporation
                 Model: Ethernet Connection (7) I219-V
            HW Address: 48:2a:e3:fa:5e:16 (Wistron InfoComm(Kunshan)Co.,Ltd.)
                   MTU: 1500 (min: 68, max: 9000)
  Queue Length (Tx/Rx): 1/1
      Auto negotiation: yes
                 Speed: 1Gbps
                Duplex: full
                  Port: tp
               Address: 10.0.1.130
                        fe80::4a2a:e3ff:fe56:2a55
               Gateway: 10.0.1.1 (NETGEAR)
```

### Logs

Debug udev logs, if needed.

```shell
SYSTEMD_LOG_LEVEL=debug udevadm test /sys/class/net/enp0s31f6
```
Output attached: [udevadm.test.debug.txt](https://github.com/systemd/systemd/files/4259683/udevadm.test.debug.txt)

## Possible Workarounds

These probably do not work in all cases, but they do in my current use cases.

```shell
cat<<EOF>/etc/systemd/network/00-ether-test.network
[Match]
Name=e*
[Network]
DHCP=ipv4
EOF
```

_Or_

```shell
cat<<EOF>/etc/systemd/network/00-ether-test.network
[Match]
Property="ID_PCI_SUBCLASS_FROM_DATABASE=Ethernet controller"
[Network]
DHCP=ipv4
EOF
```

## Other thoughts

Is it possible to use whatever `networkctl status` is using to determine `Type: ether` in addition to the `udev(7)` property `DEVTYPE`?

https://github.com/systemd/systemd/issues/3998#issuecomment-251674635 mentions possible solutions.

### PR DESCRIPTION
**Pull request by yuwata**
This makes Type= can match the type shown in networkctl.

Closes #14952.

### CODE REVIEW

**Comment by Neurognostic** [comment]
It would be possible to do https://github.com/systemd/systemd/issues/3998 after this is merged.

**Comment by Neurognostic** [comment]
I think the man page would need to be updated to explain how values other than `DEVTYPE` can be obtained/used.

**Comment by yuwata** [comment]
The man page is slightly updated.

**Comment by Neurognostic** [comment]
> The man page is slightly updated.
Does the change effect all three configuration file types (i.e. `systemd.network(5)`, `systemd.link(5)`, `systemd.netdev(5)`)?

**Comment by yuwata** [comment]
Yes. (.netdev does not have [Match] section, though.)

File: src/libsystemd/sd-netlink/netlink-util.c [hunk 1]
```diff
@@ -203,6 +203,29 @@ int rtnl_resolve_link_alternative_name(sd_netlink **rtnl, const char *name) {
...
+                if (r < 0)
+                        return r;
+        }
+
+        r = sd_rtnl_message_new_link(*rtnl, &message, RTM_GETLINK, ifindex);
+        if (r < 0)
+                return r;
+
+        r = sd_netlink_call(*rtnl, message, 0, &reply);
+        if (r == -EINVAL)
```
Comment (ssahani) [inline src/libsystemd/sd-netlink/netlink-util.c:221]: It's a sync call. But processed while networkd starts or device appears .  `Link` has the iftype information already.

Comment (yuwata) [reply src/libsystemd/sd-netlink/netlink-util.c:221]: This is called by udevd.

Comment (ssahani) [reply src/libsystemd/sd-netlink/netlink-util.c:221]: hmm . Ok.

**Comment by ssahani** [comment]
LGTM. Let's CI paases.

**Comment by yuwata** [comment]
@keszybz I set v245 milestone for this PR. But if you think it is not appropriate, then please drop this from the milestone.

File: src/udev/net/link-config.c [hunk 2]
```diff
@@ -236,9 +236,10 @@ bool link_config_should_reload(link_config_ctx *ctx) {
 
 int link_config_get(link_config_ctx *ctx, sd_device *device, link_config **ret) {
         struct ether_addr permanent_mac = {};
+        unsigned short iftype = 0;
```
Comment (keszybz) [inline src/udev/net/link-config.c:239]: This doesn't need to be initialized.

**Pull request merged**